presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PropertyDerivations.java

```diff
@@ -826,6 +838,82 @@ private Global deriveGlobalProperties(TableLayout layout, Map<ColumnHandle, Vari
         return inputToOutput;
     }
 
+    private static Optional<RowExpression> rewriteToCoalesce(
+            TypeManager typeManager,
+            ImmutableBiMap<VariableReferenceExpression, VariableReferenceExpression> joinMap,
+            RowExpression column)
+    {
+        checkArgument(column instanceof VariableReferenceExpression || column instanceof ConstantExpression, "Expect input partition to be variables or constants");
+        if (column instanceof VariableReferenceExpression) {
+            VariableReferenceExpression variable = (VariableReferenceExpression) column;
+            ImmutableBiMap map;
+            if (!joinMap.containsKey(variable) && !joinMap.containsValue(variable)) {
+                return Optional.empty();
+            }
+            if (joinMap.containsKey(variable)) {
+                map = joinMap;
+            }
+            else {
+                map = joinMap.inverse();
+            }
+            return Optional.of(new SpecialFormExpression(
+                    COALESCE,
+                    typeManager.getCommonSuperType(variable.getType(), ((VariableReferenceExpression) map.get(variable)).getType()).get(),
+                    ImmutableList.of(column, (RowExpression) map.get(variable))));
+        }
+        else {
+            return Optional.of(column);
+        }
+    }
+
+    private static Optional<RowExpression> rewriteExpression(
+            Map<VariableReferenceExpression, RowExpression> assignments,
+            TypeProvider types,
+            Map<VariableReferenceExpression, VariableReferenceExpression> identities,
+            RowExpression expression)
+    {
+        checkArgument(expression instanceof VariableReferenceExpression || (expression instanceof SpecialFormExpression && ((SpecialFormExpression) expression).getForm().equals(COALESCE)), "The rewrite can only handle CoalesceExpression");
+        if (expression instanceof VariableReferenceExpression) {
+            return Optional.ofNullable(identities.get(expression));
+        }
+
+        Set<RowExpression> arguments = ImmutableSet.copyOf(((SpecialFormExpression) expression).getArguments());
+        checkArgument(arguments.stream().allMatch(VariableReferenceExpression.class::isInstance), "Expect arguments of COALESCE to be VariableReferenceExpression");
+        // We are using the property that the result of coalesce from full outer join keys would not be null despite of the order
+        // of the arguments. Thus we extract and compare the variables of the COALESCE as a set rather than compare COALESCE directly.
+        for (Map.Entry<VariableReferenceExpression, RowExpression> entry : assignments.entrySet()) {
+            if (isExpression(entry.getValue())) {
```

Comment:
Curious: Under what circumstance the `RowExpression` here will be an `OriginalExpression`?

Reply:
Depending on the progress of the Expression to RowExpression refactoring. Basically if the translation happens before this we should only expect `RowExpression`, otherwise we should only have `OriginalExpression`. Handling both here just in case. We can remove the expression part once the refactor is done (when we delete `OriginalExpression`).